File: 백준/Gold/17472. 다리 만들기 2/다리 만들기 2.cc
#include<iostream>
#include<algorithm>
#include<queue>
#include<vector>
#include<cstring>
#include<tuple>
using namespace std;

int N, M, result;;
int Map[11][11];
int vis[11][11];
int parent[7];// 2<=섬<=6

int dx[]{ 1,-1,0,0 };
int dy[]{ 0,0,1,-1 };

vector<tuple<int, int, int> > bridge;
vector<pair<int, int> > v;

int find(int a)
{
	if (parent[a] == a)
		return a;
	return parent[a] = find(parent[a]);
}

void Union(int a, int b)
{
	a = find(a);
	b = find(b);
	if (a != b)
	{
		if (a < b)
			parent[b] = a;
		else
			parent[a] = b;
	}

}

bool sameParent(int a, int b)
{
	a = find(a);
	b = find(b);
	if (a != b)
		return false;
	return true;

}
//각 섬에 번호 매기기
void count_island(int x, int y, int num) {
	queue<pair<int, int>> q;
	q.push({ x,y });
	vis[x][y] = 1;
	Map[x][y] = num;

	while (!q.empty()) {
		auto [cx,cy] = q.front();
		q.pop();

		for (auto i = 0; i < 4; ++i) {
			int nx = cx + dx[i];
			int ny = cy + dy[i];

			if (nx < 0 || nx >= N || ny < 0 || ny >= M)continue;
			if (!vis[nx][ny] && Map[nx][ny] == -1) {
				vis[nx][ny] = 1;
				Map[nx][ny] = num;
				q.push({ nx,ny });
			}
		}
	}
}

//입력받은 방향대로 다리 만들기
void make_bridge(int x, int y, int dir) {
	int len = 0;
	int start_point = Map[x][y];

	while (1) {
		int nx = x + dx[dir];
		int ny = y + dy[dir];

		if (!(nx < 0 || ny < 0 || nx >= N || ny >= M)) {
			if (Map[nx][ny] == 0) {
				x = nx;
				y = ny;
				len++;
			}
			//길이가 2 이상이어야해서 조건 추가
			else if (len >= 2 && start_point != Map[nx][ny]) {
				bridge.push_back({ len,start_point,Map[nx][ny] });
				break;
			}
			else
				break;
			
		}
		else
			break;
		

	}
}


int main(void) {
	cin.tie(NULL)->sync_with_stdio(false);

	cin >> N >> M;
	for (int i = 0; i < N; ++i) {
		for (int j = 0; j < M; ++j) {
			cin >> Map[i][j];
			if (Map[i][j] == 1) {
				Map[i][j] = -1;
				v.push_back({ i,j });
			}
		}
	}

	int cnt = 1;
	for (int i = 0; i < N; ++i) {
		for (int j = 0; j < M; ++j) {
			if (Map[i][j] == -1 && !vis[i][j]) {
				count_island(i, j, cnt);
				cnt++;
			}
		}
	}

	for (auto& [x, y] : v) {
		for (int i = 0; i < 4; ++i)
			make_bridge(x, y, i);
	}

	sort(bridge.begin(), bridge.end());

	for (int i = 1; i <= cnt; ++i) // union-find 알고리즘을 위해 초기화
		parent[i] = i;

	for (auto& [len, start, dest] : bridge) {
		if (!sameParent(start, dest)) {
			Union(start, dest);
			result += len;
		}
	}

	for (int i = 1; i < cnt; ++i) {
		if (find(i) != 1) {
			cout << -1;
			return 0;
		}
	}

	cout << result;
	return 0;
}

File: 백준/Gold/17472. 다리 만들기 2/다리 만들기 2.java
import java.io.*;
import java.util.*;

public class Main {
    static int N, M;
    static int[][] board;
    static boolean[][] vis;
    static PriorityQueue<Node> pq = new PriorityQueue<>();
    static int[] parent, rank;
    static int[] dx = {0, 1, 0, -1};
    static int[] dy = {1, 0, -1, 0};

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        N = Integer.parseInt(st.nextToken());
        M = Integer.parseInt(st.nextToken());

        board = new int[N][M];
        vis = new boolean[N][M];

        for (int i = 0; i < N; i++) {
            st = new StringTokenizer(br.readLine());

            for (int j = 0; j < M; j++) {
                board[i][j] = Integer.parseInt(st.nextToken());
                if (board[i][j] == 1) board[i][j] = -1;
            }
        }

        int count = 1;
        for (int i = 0; i < N; i++) {
            for (int j = 0; j < M; j++) {
                if (board[i][j] == -1 && !vis[i][j]) {
                    maskMap(count, i, j);
                    count++;
                }
            }
        }

        for (int i = 0; i < N; i++) {
            for (int j = 0; j < M; j++) {
                if (board[i][j] > 0) {
                    for (int d = 0; d < 4; d++) {
                        findBridge(i, j, d, board[i][j]);
                    }
                }
            }
        }

        parent = new int[count];
        rank = new int[count];
        for (int i = 1; i < count; i++) {
            parent[i] = i;
        }

        int cnt = 0;
        int ans = 0;
        while (!pq.isEmpty()) {
            Node cur = pq.poll();

            if (union(cur.from, cur.to)) {
                ans += cur.cost;
                cnt++;
            }
        }

        if (cnt == count - 2) {
            System.out.println(ans);
        }else{
            System.out.println(-1);
        }

    }

    //맵에 섬별로 마스킹 하기
    static void maskMap(int num, int x, int y) {
        Queue<int[]> q = new ArrayDeque<>();
        q.offer(new int[]{x, y});

        while (!q.isEmpty()) {
            int[] cur = q.poll();
            int cx = cur[0];
            int cy = cur[1];
            board[cx][cy] = num;
            vis[cx][cy] = true;

            for (int i = 0; i < 4; i++) {
                int nx = cx + dx[i];
                int ny = cy + dy[i];

                if (nx < 0 || ny < 0 || nx >= N || ny >= M) continue;
                if (vis[nx][ny]) continue;
                if (board[nx][ny] != 0 && !vis[nx][ny]) {
                    board[nx][ny] = num;
                    vis[nx][ny] = true;
                    q.offer(new int[]{nx, ny});
                }
            }
        }
    }

    static void findBridge(int x, int y, int dir, int start) {
        int dist = 0;
        int nx = x;
        int ny = y;

        while (true) {
            nx += dx[dir];
            ny += dy[dir];

            if (nx < 0 || ny < 0 || nx >= N || ny >= M) break;
            if (board[nx][ny] == start) break;

            if (board[nx][ny] == 0) {
                dist++;
            } else {
                if (dist >= 2) {
                    pq.offer(new Node(start, board[nx][ny], dist));
                }
                break;
            }
        }
    }

    static int find(int x) {
        if (parent[x] == x) return x;
        return parent[x] = find(parent[x]);
    }

    static boolean union(int a, int b) {
        int rootA = find(a);
        int rootB = find(b);

        if (rootA == rootB) return false;
        else {
            if (rank[rootA] < rank[rootB]) {
                parent[rootA] = rootB;
            } else if (rank[rootB] < rank[rootA]) {
                parent[rootB] = rootA;
            } else {
                parent[rootB] = rootA;
                rank[rootA]++;
            }
            return true;
        }
    }

    static class Node implements Comparable<Node> {
        int from, to;
        int cost;

        public Node(int from, int to, int cost) {
            this.from = from;
            this.to = to;
            this.cost = cost;
        }

        @Override
        public int compareTo(Node o) {
            return this.cost - o.cost;
        }
    }
}
